core-tools.markdown

```diff
@@ -0,0 +1,345 @@
+without warnings.
+
+`-Wall` emits many useful warnings that can easily be bugs,
+including warnings about name shadowing, unused variables, and more.
+
+Note that `-Wall` does not stop GHC from compiling your program.
+If this is something you'd like to do, specifying the flag `-Werror`
+will convert warnings to errors and will halt compilation when a warning is emitted.
+
+#### Optimisations
+
+Another very useful flag is `-O`. This flag asks GHC to compile the program with optimisations and code improvements at the cost of taking longer to compile.
+This will often make the program run much faster and perform more aggressive optimisations such as inlining and specialisation.
+For even more optimisations that may take significantly longer, `-O2` is also available.
+
+## An interactive environment
+
+GHC provides an interactive environment in a form of a
+Read-Evaluate-Print Loop (REPL) called GHCi.
+To enter the environment run the program `ghci`.
+
+```sh
+➜ ghci
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> 
+```
+
+It provides an interactive prompt where Haskell expressions can be written and
+evaluated.
+
+For example:
+
+```sh
+ghci> 1 + 1
+2
+ghci> putStrLn "Hello, world!"
+Hello, world!
+```
+
+We can define new names:
+
+```sh
+ghci> double x = x + x
+ghci> double 2
+4
+```
+
+We can import Haskell source files using the `:load` command:
+
+```sh
+ghci> :load hello.hs
+[1 of 1] Compiling Main             ( hello.hs, interpreted )
+Ok, one module loaded.
+ghci> main
+Hello, Haskell!
+```
+
+As well as import library modules:
+
+```sh
+ghci> import Data.Bits
+ghci> shiftL 32 1
+64
+ghci> clearBit 33 0
+32
+```
+
+We can even ask what the type of an expression is using the `:type` command:
+
+```sh
+λ> :type putStrLn
+putStrLn :: String -> IO ()
+```
+
+To exit `ghci`, use the `:quit` command (or `:q` for short)
+
+```sh
+ghci> :quit
+Leaving GHCi.
+```
+
+A more thorough introduction to GHCi can be found in the [GHC user guide](https://downloads.haskell.org/~ghc/9.0.1/docs/html/users_guide/ghci.html).
+
+### Using external packages in ghci
+
+By default, ghci can only load and use packages that are [included with the GHC installation](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/9.0.1-notes.html#included-libraries).
+
+However, users of [cabal-install](https://www.haskell.org/cabal) and [stack](http://haskellstack.org) can download and load external packages very easily using the following commands:
+
+Cabal-install:
+
+```sh
+cabal repl --build-depends random
+```
+
+Stack:
+
+```sh
+stack exec --package random -- ghci
+```
+
+And the modules of the relevant packages will be available for import:
+
+```sh
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+λ> import System.Random
+λ> randomRIO (1, 10)
+1
+λ> randomRIO (1, 10)
+8
+λ> randomRIO (1, 10)
+5
+```
+
+Stack users can also use this feature with `runghc` and `ghc` by replacing `ghci` in the command above.
```

Comment:
For symmetry I'd add that `cabal --lib --package-env . install random` will generate an environment file that will make `random` visible for GHC / runghc calls in the current directory. Alternatively, I'd remove this topic altogether.

Reply:
I'd very much like to add such a command to the tutorial, but I tried running said command but ran into an error:

```sh
➜  cabal --lib --package-env . install random
cabal: unrecognised command: . (try --help)
```

Reply:
I'm sorry, should have checked: `install` should go first:

```
cabal install --lib random --package-env .
```

Reply:
Thanks! I've added this option and replaced the example with an example that requires two packages.